### PR TITLE
attempt to replace nose with pytest

### PR DESCRIPTION
**Pull request by emmanuelle**

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
💃 

Maybe we need to remove `nosetests` from some requirements files also? probably not a blocker.

**Comment by emmanuelle** [comment]
Apparently nose tools are used inside the test files:
```
$ grep -r "from nose" *
packages/python/chart-studio/chart_studio/tests/test_plot_ly/test_stream/test_stream.py~:from nose.plugins.attrib import attr
packages/python/chart-studio/chart_studio/tests/test_plot_ly/test_stream/test_stream.py~:from nose.tools import nottest
packages/python/chart-studio/chart_studio/tests/test_plot_ly/test_get_figure/test_get_figure.py:from nose.plugins.attrib import attr
packages/python/chart-studio/chart_studio/tests/test_plot_ly/test_get_requests/test_get_requests.py:from nose.plugins.attrib import attr
packages/python/chart-studio/chart_studio/tests/test_plot_ly/test_plotly/test_plot.py:from nose.plugins.attrib import attr
packages/python/chart-studio/chart_studio/tests/test_plot_ly/test_image/test_image.py:from nose.plugins.attrib import attr
packages/python/chart-studio/chart_studio/tests/test_plot_ly/test_meta/test_meta.py:from nose.plugins.attrib import attr
packages/python/chart-studio/chart_studio/tests/test_plot_ly/test_file/test_file.py:from nose.plugins.attrib import attr
packages/python/chart-studio/chart_studio/tests/test_plot_ly/test_grid/test_grid.py:from nose.plugins.attrib import attr
packages/python/chart-studio/chart_studio/tests/test_core/test_tools/test_get_embed.py:from nose.tools import raises
packages/python/chart-studio/chart_studio/tests/test_optional/test_matplotlylib/test_plot_mpl.py:from nose.plugins.attrib import attr
packages/python/chart-studio/chart_studio/tests/test_optional/test_matplotlylib/test_plot_mpl.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_figure_widget_backend/test_validate_no_frames.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_offline/test_offline.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_core/test_figure_messages/test_move_delete_traces.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_figure_messages/test_on_change.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_colors/test_colors.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_annotations.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_append_trace.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_properties_validated.py~:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_properties_validated.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_constructor.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_template.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_frames.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_layout_subplots.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_data.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_error_bars.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_figure_properties.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_core/test_graph_objs/test_scatter.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_optional/test_offline/test_offline.py:from nose.tools import raises
packages/python/plotly/plotly/tests/test_optional/test_offline/test_offline.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_optional/test_utils/test_utils.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_optional/test_matplotlylib/test_annotations.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_optional/test_matplotlylib/test_axis_scales.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_optional/test_matplotlylib/test_date_times.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_optional/test_matplotlylib/test_bars.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_optional/test_matplotlylib/test_subplots.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_optional/test_matplotlylib/test_data.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_optional/test_matplotlylib/test_lines.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/tests/test_optional/test_matplotlylib/test_scatter.py:from nose.plugins.attrib import attr
packages/python/plotly/plotly/matplotlylib/mplexporter/tests/test_basic.py:from nose.plugins.skip import SkipTest
```
so this one needs more work. I'd prefer getting rid of nose as a whole instead of just switching to running pytest but keeping to importing nose. It doesn't matter whether this PR is merged before of after the patch release.

**Comment by emmanuelle** [comment]
Ha ha can't believe the CI was green on the first try! Happy reviewing... only 39 files :-p

File: packages/python/chart-studio/chart_studio/tests/test_optional/test_matplotlylib/test_plot_mpl.py [hunk 1]
```diff
@@ -7,40 +7,37 @@
...
 import _plotly_utils.exceptions
 from plotly import optional_imports
 from chart_studio.plotly import plotly as py
 from unittest import TestCase
+import pytest
 
 matplotlylib = optional_imports.get_module("plotly.matplotlylib")
 
 if matplotlylib:
     import matplotlib.pyplot as plt
```
Comment (emmanuelle) [inline packages/python/chart-studio/chart_studio/tests/test_optional/test_matplotlylib/test_plot_mpl.py]:
```suggestion
@pytest.mark.matplotlib
```

**Comment by emmanuelle** [comment]
I removed the "slow" attributes because we're not using them anywhere in the CI but this can be discussed of course. Sometimes it's good to remove unused code :-).

**Comment by emmanuelle** [comment]
@nicolaskruchten you gave a :dancer: but this was before I made a number of changes to remove completely the nose dependency, can you please confirm whether I can merge?

**Comment by nicolaskruchten** [comment]
One last thing: we should remove references to `nose` from https://github.com/plotly/plotly.py/blob/master/contributing.md and any other non-code documentation please :) 

After that, 💃

File: contributing.md [hunk 2]
```diff
@@ -128,34 +128,34 @@ classes based on the new schema.
...
-nosetests -w  packages/python/plotly/plotly/tests/test_core/
+pytest  packages/python/plotly/plotly/tests/test_core/
 ```
 
 ... or even more granular by running something like:
 
 ```bash
-nosetests plotly/tests/test_plotly/test_plot.py
+pytest plotly/tests/test_plotly/test_plot.py
 ```
```
Comment (nicolaskruchten) [inline contributing.md]: Oh can we add here the incantation for running a single test? I think it’s with a colon? I always forget it!

**Pull request merged**